conjure-http/src/server.rs

```diff
@@ -300,3 +327,334 @@ impl<'a> RequestContext<'a> {
         self.response_extensions
     }
 }
+
+enum MaybeBorrowed<'a, T> {
+    Borrowed(&'a T),
+    Owned(T),
+}
+
+impl<T> Deref for MaybeBorrowed<'_, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        match self {
+            MaybeBorrowed::Borrowed(v) => v,
+            MaybeBorrowed::Owned(v) => v,
+        }
+    }
+}
+
+/// A trait implemented by request body deserializers used by custom Conjure server trait
+/// implementations.
+pub trait DeserializeRequest<T, R> {
+    /// Deserializes the request body.
+    fn deserialize(headers: &HeaderMap, body: R) -> Result<T, Error>;
+}
+
+/// A trait implemented by response deserializers used by custom async Conjure server trait
+/// implementations.
+#[async_trait]
+pub trait AsyncDeserializeRequest<T, R> {
+    /// Deserializes the request body.
+    async fn deserialize(headers: &HeaderMap, body: R) -> Result<T, Error>
+    where
+        R: 'async_trait;
+}
+
+/// A request deserializer which acts as a Conjure-generated endpoint would.
+pub enum ConjureRequestDeserializer {}
+
+impl ConjureRequestDeserializer {
+    fn check_content_type(headers: &HeaderMap) -> Result<(), Error> {
+        if headers.get(CONTENT_TYPE) != Some(&APPLICATION_JSON) {
+            return Err(Error::service_safe(
+                "invalid request Content-Type",
+                InvalidArgument::new(),
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+impl<T, R> DeserializeRequest<T, R> for ConjureRequestDeserializer
+where
+    T: DeserializeOwned,
+    R: Iterator<Item = Result<Bytes, Error>>,
+{
+    fn deserialize(headers: &HeaderMap, body: R) -> Result<T, Error> {
+        Self::check_content_type(headers)?;
+        let buf = private::read_body(body, Some(SERIALIZABLE_REQUEST_SIZE_LIMIT))?;
+        json::server_from_slice(&buf).map_err(|e| Error::service(e, InvalidArgument::new()))
+    }
+}
+
+#[async_trait]
+impl<T, R> AsyncDeserializeRequest<T, R> for ConjureRequestDeserializer
+where
+    T: DeserializeOwned,
+    R: Stream<Item = Result<Bytes, Error>> + Send,
+{
+    async fn deserialize(headers: &HeaderMap, body: R) -> Result<T, Error>
+    where
+        R: 'async_trait,
+    {
+        Self::check_content_type(headers)?;
+        let buf = private::async_read_body(body, Some(SERIALIZABLE_REQUEST_SIZE_LIMIT)).await?;
+        json::server_from_slice(&buf).map_err(|e| Error::service(e, InvalidArgument::new()))
+    }
+}
+
+/// A trait implemented by response serializers used by custom Conjure server trait implementations.
+pub trait SerializeResponse<T, W> {
+    /// Serializes the response.
+    fn serialize(request_headers: &HeaderMap, value: T)
```

Comment:
I wonder if it may be worth passing in a struct that has the request headers and the values instead of passing them in as separate args? That way if there are other fields we care about in the future we can add those without making a breaking change to trait implementations

Reply:
I have this change partially implemented, but it's going to be a bit large. I think it probably makes sense to split out to a follow-up PR.

Reply:
https://github.com/palantir/conjure-rust/pull/258